public/app/features/alerting/unified/hooks/usePrometheusConsistencyCheck.ts

```diff
@@ -43,6 +52,172 @@ function useMatchingPromRuleExists() {
+        ruleSourceName: dsFeatures.name,
+        namespace: namespace,
+        groupName: ruleIdentifier.groupName,
+      };
+      const rulerParams = {
+        namespace,
+        group: ruleIdentifier.groupName,
+        rulerConfig: dsFeatures.rulerConfig,
+      };
+
+      const [promResponse, rulerResponse] = await Promise.allSettled([
+        fetchPrometheusRuleGroups(promQueryParams),
+        fetchRuleGroup(rulerParams),
+      ]);
+
+      if (promResponse.status === 'rejected' && rulerResponse.status === 'rejected') {
+        // This means both requests failed. We can't determine if the state is consistent or not
+        // and most probably mean there is a connectivity issue with the datasource
+        // Let's return true so the UI is not disruptive for the user, but log an error to investigate how often this happens
+        logError(
+          new Error('Error fetching Prometheus and Ruler rule groups', {
+            cause: [promResponse.reason, rulerResponse.reason],
+          })
+        );
+        return true;
+      }
+
+      if (promResponse.status === 'rejected' && rulerResponse.status === 'fulfilled') {
+        // This means Prometheus request error. It shouldn't reject even if there are no groups
+        // matching the query params
+        // Let's return true so the UI is not disruptive for the user, but log an error to investigate how often this happens
+        logError(new Error('Error fetching Prometheus rule groups', { cause: promResponse.reason }));
+        return true;
+      }
+
+      if (rulerResponse.status === 'rejected' && promResponse.status === 'fulfilled') {
+        // We assume the group no longer exists in the ruler
+        // The state is consistent if
+        const promGroups = promResponse.value.data?.flatMap((ns) => ns.groups);
+        return !promGroups?.some((g) => g.name === ruleIdentifier.groupName);
+      }
+
+      if (promResponse.status === 'fulfilled' && rulerResponse.status === 'fulfilled') {
+        const promGroup = promResponse.value.data
+          ?.flatMap((ns) => ns.groups)
+          .find((g) => g.name === ruleIdentifier.groupName);
+        const rulerGroup = rulerResponse.value.data;
+
+        if (promGroup && rulerGroup) {
+          const rulesCountMatches = promGroup.rules.length === rulerGroup.rules.length;
+          if (!rulesCountMatches) {
+            return false;
+          }
+
+          const promRuleNames = promGroup.rules.map((r) => r.name);
+          const rulerRuleNames = rulerGroup.rules.map(getRuleName);
+
+          for (const [promName, rulerName] of zip(promRuleNames, rulerRuleNames)) {
+            if (promName !== rulerName) {
+              return false;
+            }
+          }
+
+          return true;
+        }
+      }
+
+      return false;
+    },
+    [discoverDsFeatures, fetchPrometheusRuleGroups, fetchRuleGroup]
+  );
+
+  return { isGroupInSync };
+}
+
+export function useRuleGroupConsistencyCheck() {
```

Comment:
Discussed offline, but perhaps worth looking in to a similar pattern I've used in the AM settings page for waiting on remote state by invalidating cache until some condition was met;

https://github.com/grafana/grafana/blob/07b069db83b8a2415a7e2125736aadbf3b6cc1d2/public/app/features/alerting/unified/components/settings/SettingsContext.tsx#L172-L187

Reply:
not a blocker; so I'm marking this as resolved ;)